R/cbc_solve.R

```diff
@@ -37,42 +39,64 @@ cbc_solve <- function(obj,
                       col_ub = rep.int(Inf, length(obj)),
                       is_integer = rep.int(FALSE, length(obj)),
                       max = FALSE, cbc_args = list()) {
-  stopifnot(is.numeric(obj))
-  stopifnot(is.numeric(row_ub))
+  assert_that(is.numeric(obj), is.numeric(row_ub))
   is_sparse_matrix <- inherits(mat, "dgTMatrix")
   if (!is_sparse_matrix) {
     mat <- methods::as(Matrix::Matrix(mat), "TsparseMatrix")
   }
-  stopifnot(length(obj) == ncol(mat))
-  stopifnot(length(col_lb) == ncol(mat))
-  stopifnot(length(col_ub) == ncol(mat))
-  stopifnot(length(is_integer) == ncol(is_integer))
-  stopifnot(length(row_lb) == nrow(mat))
-  stopifnot(length(row_ub) == nrow(mat))
-  stopifnot(length(is_integer) == ncol(mat))
-  cbc_args <- unlist(lapply(cbc_args, as.character),
-                                  use.names = TRUE)
-  if (!is.character(cbc_args)) {
+
+  assert_that(
+    length(obj) == ncol(mat),
+    length(col_lb) == ncol(mat),
+    length(col_ub) == ncol(mat),
+    length(is_integer) == ncol(mat),
+    length(row_lb) == nrow(mat),
+    length(row_ub) == nrow(mat)
+  )
+
+  cbc_args <- do.call(prepare_cbc_args, cbc_args)
+
+  result <- cpp_cbc_solve(
+    obj = obj,
+    isMaximization = max,
+    rowIndices = mat@i,
+    colIndices = mat@j,
+    elements = mat@x,
+    integerIndices = which(is_integer) - 1,
+    colLower = col_lb,
+    colUpper = col_ub,
+    rowLower = row_lb,
+    rowUpper = row_ub,
+    arguments = cbc_args
+  )
+  structure(result, class = "rcbc_milp_result")
+}
+
+#' Prepares list of arguments into format accepted by cbc
+#'
+#' @noRd
+#' @param ... parameters for the cbc optimisation
+prepare_cbc_args <- function(...) {
```

Comment:
Why `...` and not just a single argument that is a list? Because in line 57 you convert the list to multiple arguments and then in line 80 back to a list. You could pass `cbc_args` directly.

Reply:
I could, but it makes it easier to test and and reason (for me).

Reply:
It just makes it a bit harder to program against that interface and simply passing a named list. But it is also just an internal function, to why not.